yjit/src/codegen.rs

```diff
@@ -3700,116 +3703,102 @@ fn gen_send_cfunc(jit: &JITState, ctx: &Context, ci: * const rb_callinfo, cme: *
-        RUBY_ASSERT(false);
-        break;
-
-      case SHAPE_DEFAULT:
-        mov(cb, REG0, const_ptr_opnd(target0));
-        mov(cb, member_opnd(REG_CFP, rb_control_frame_t, jit_return), REG0);
-        break;
+    match shape {
+        BranchShape::Next0 | BranchShape::Next1 => unreachable!(),
+        BranchShape::Default => {
+            mov(cb, REG0, const_ptr_opnd(target0.raw_ptr()));
+            mov(cb, mem_opnd(64, REG_CFP, RUBY_OFFSET_CFP_JIT_RETURN), REG0);
+        }
     }
 }
 
-// Returns whether the iseq only needs positional (lead) argument setup.
-static bool
-iseq_lead_only_arg_setup_p(const rb_iseq_t *iseq)
-{
-    // When iseq->body->local_iseq == iseq, setup_parameters_complex()
-    // doesn't do anything to setup the block parameter.
-    bool takes_block = iseq->body->param.flags.has_block;
-    return (!takes_block || iseq->body->local_iseq == iseq) &&
-        iseq->body->param.flags.has_opt          == false &&
-        iseq->body->param.flags.has_rest         == false &&
-        iseq->body->param.flags.has_post         == false &&
-        iseq->body->param.flags.has_kw           == false &&
-        iseq->body->param.flags.has_kwrest       == false &&
-        iseq->body->param.flags.accepts_no_kwarg == false;
-}
-
-bool rb_iseq_only_optparam_p(const rb_iseq_t *iseq);
-bool rb_iseq_only_kwparam_p(const rb_iseq_t *iseq);
-
 // If true, the iseq is leaf and it can be replaced by a single C call.
-static bool
-rb_leaf_invokebuiltin_iseq_p(const rb_iseq_t *iseq)
+fn rb_leaf_invokebuiltin_iseq_p(iseq: IseqPtr) -> bool
 {
-    unsigned int invokebuiltin_len = insn_len(BIN(opt_invokebuiltin_delegate_leave));
-    unsigned int leave_len = insn_len(BIN(leave));
+    use std::os::raw::{c_void};
 
-    return (iseq->body->iseq_size == (invokebuiltin_len + leave_len) &&
-        rb_vm_insn_addr2opcode((void *)iseq->body->iseq_encoded[0]) == BIN(opt_invokebuiltin_delegate_leave) &&
-        rb_vm_insn_addr2opcode((void *)iseq->body->iseq_encoded[invokebuiltin_len]) == BIN(leave) &&
-        iseq->body->builtin_inline_p
-    );
+    let invokebuiltin_len = insn_len(OP_OPT_INVOKEBUILTIN_DELEGATE_LEAVE);
+    let leave_len = insn_len(OP_LEAVE);
+
+    unsafe {
+        let enc_size = get_iseq_encoded_size(iseq);
+        let enc = get_iseq_body_iseq_encoded(iseq);
+        let bip = get_iseq_body_builtin_inline_p(iseq);
+
+        let first_opcode_raw = (*enc).as_u64() as *const u8;
+        let first_opcode = rb_vm_insn_addr2opcode(first_opcode_raw as *const c_void);
+        let offs = invokebuiltin_len as isize * SIZEOF_VALUE as isize;
+        let second_opcode_raw = (*(enc.offset(offs))).as_u64() as *const u8;
+        let second_opcode = rb_vm_insn_addr2opcode(second_opcode_raw as *const c_void);
+
+        return enc_size == (invokebuiltin_len + leave_len) &&
+            first_opcode == (OP_OPT_INVOKEBUILTIN_DELEGATE_LEAVE as i32) &&
+            second_opcode == (OP_LEAVE as i32) && bip;
+    };
  }
```

Comment:
I think we'll want to keep this bit of code in C, as you did with the other predicate above.

Reply:
Fair. These are fairly messy in Rust. I'll switch them over.